ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/SnapshotTrieLogManager.java

```diff
@@ -62,51 +63,75 @@ public void addCachedLayer(
-        worldStateRootHash::toHexString);
+        worldStateRootHash::toShortHexString);
+
+    BonsaiSnapshotWorldState snapshot =
+        BonsaiSnapshotWorldState.create(worldStateArchive, worldStateStorage);
+
     cachedWorldStatesByHash.put(
         blockHeader.getHash(),
-        new CachedSnapshotWorldState(
-            () ->
-                worldStateArchive
-                    .getMutableSnapshot(blockHeader.getHash())
-                    .map(BonsaiSnapshotWorldState.class::cast)
-                    .orElse(null),
-            trieLog,
-            blockHeader.getNumber()));
+        new CachedSnapshotWorldState(snapshot, trieLog, blockHeader.getNumber()));
   }
 
   @Override
-  public Optional<MutableWorldState> getBonsaiCachedWorldState(final Hash blockHash) {
+  public void updateCachedLayers(final Hash blockParentHash, final Hash blockHash) {
+    // no-op.
+  }
+
+  @Override
+  public synchronized Optional<MutableWorldState> getBonsaiCachedWorldState(final Hash blockHash) {
     if (cachedWorldStatesByHash.containsKey(blockHash)) {
       return Optional.ofNullable(cachedWorldStatesByHash.get(blockHash))
-          .map(CachedSnapshotWorldState::getMutableWorldState)
+          .map(CachedWorldState::getMutableWorldState)
           .map(MutableWorldState::copy);
     }
     return Optional.empty();
   }
 
   @Override
-  public void updateCachedLayers(final Hash blockParentHash, final Hash blockHash) {
-    // fetch the snapshot supplier as soon as its block has been added:
-    Optional.ofNullable(cachedWorldStatesByHash.get(blockHash))
-        .ifPresent(CachedSnapshotWorldState::getMutableWorldState);
+  public synchronized void onClear() {
+    dropArchive();
+  }
+
+  @Override
+  public synchronized void onClearFlatDatabase() {
+    dropArchive();
   }
 
-  public static class CachedSnapshotWorldState implements CachedWorldState {
+  private void dropArchive() {
+    // drop all cached snapshot worldstates, they are unsafe when the db has been truncated
+    LOG.info("Key-value storage truncated, dropping cached worldstates");
+    cachedWorldStatesByHash.keySet().forEach(cachedWorldStatesByHash::remove);
+  }
 
-    final Supplier<BonsaiSnapshotWorldState> snapshot;
+  public static class CachedSnapshotWorldState
+      implements CachedWorldState<BonsaiSnapshotWorldState> {
+
+    final BonsaiSnapshotWorldState snapshot;
+    final Long worldStateSubscriberId;
     final TrieLogLayer trieLog;
     final long height;
+    final AtomicBoolean isClosed = new AtomicBoolean(false);
 
     public CachedSnapshotWorldState(
-        final Supplier<BonsaiSnapshotWorldState> snapshotSupplier,
-        final TrieLogLayer trieLog,
-        final long height) {
-      this.snapshot = Suppliers.memoize(snapshotSupplier::get);
+        final BonsaiSnapshotWorldState snapshot, final TrieLogLayer trieLog, final long height) {
+      this.worldStateSubscriberId = snapshot.getWorldStateStorage().subscribe(this);
```

Comment:
we are not already subscribe here ? https://github.com/hyperledger/besu/pull/4786/files#diff-8208cce046b9c78ec905315f4bbbc0a4813abcacd299489acc59833bbce7d35aR50

Reply:
1. SnapshotTrieLogManager - subscribes so it can drop all cached layers after a truncate event
2. BonsaiSnapshotWorldState - subscribes so it can prevent read/write through the snapshot transaction after a truncate event.

If we just had #1, a reference to a snapshot that has already been fetched from the trie log manager can segfault
if we just had #2, then we would keep a bunch of cached worldstates that will refuse to read/write

We need both